### ISSUE TITLE
WaveOut.Init intermittently fails on some devices

### ISSUE THREAD

**Issue by chikubian** [naudio/NAudio#100]
Hi Mark...

Firstly, thank you for your excellent work on NAudio.

I've just encountered a new issue with NAudio's WaveOut class which is crashing intermittently on some machines.

The exception that I am getting is:

```
NAudio.MmException: AlreadyAllocated calling waveOutOpen
   at NAudio.Wave.WaveOut.Init(IWaveProvider waveProvider)
   at MyCode.NAudioPlayer.Play()
```

I've been careful to follow your usage examples, and to ensure that all resources are disposed when I'm finished with them. And this issue occurs on the second or third attempt to play, so it's not likely to be a case of resources running out.

Here is a summary of the code that I use to play an MP3 file, loopable and with run-time adjustable volume... wrapped into my helper class called NAudioPlayer:

```
public class NAudioPlayer 
{
    private string _mediaFile; // which is set in the constructor

    . . .

    public void Play() 
    {
        _fileReader = new MediaFoundationReader(_mediaFile);
        _loopStream = new LoopStream(_fileReader);
        _loopStream.EnableLooping = _enableLooping;
        _waveOut = new WaveOut();
        _volumeProvider = new VolumeSampleProvider(_loopStream.ToSampleProvider());
        _volumeProvider.Volume = 1;   // simplified for example
        _waveOut.Init(_volumeProvider);     // <== this is the line that is failing
        _waveOut.Play();
    }

    public void Dispose() 
    {
        _waveOut.Stop();
        _waveOut.Dispose();
        _loopStream.Dispose();
        _fileReader.Dispose();
    }
}
```

The issue can be reproduced with the following steps carried out manually in our GUI but depicted here as just code:

```
var player1 = new NAudioPlayer("Test.mp3");
player1.Play();

// ... wait a while then dispose while it's still playing...
sleep(a_short_time);
player1.Dispose();

// ... wait a second or so, then start two concurrent playbacks a short time apart...
var player2 = new NAudioPlayer("Test.mp3");
player2.Play();
sleep(a_short_time);
var player3 = new NAudioPlayer("Test.mp3");
player3.Play();    //  <=== KABOOM!!! the exception occurs here

```

If I substitute WaveOut with DirectSoundOut, it becomes 100% reliable at allocating the wave player, however playback becomes a bit glitchy, with clicks, jumps and skips that occur every second or two throughout the audio file being played. From what I've read in your documentation, WaveOut seems to be the safest option. 

The audio driver being used is the latest Realtek driver on Windows 7.

Any advice or suggestions for a fix would be greatly appreciated.

Kind regards,
Dave.

**Comment by markheath** [comment]
I've had problems with Realtek in the past, but upgrading drivers has usually fixed it. See if using WaveOutEvent works better

**Comment by chikubian** [comment]
Thanks for your reply, Mark. I've just tried WaveOutEvent and am having the exact same issue.

**Comment by markheath** [comment]
how many concurrent instances of WaveOut do you have on the same device? One option is to open the soundcard only once and use MixingSampleProvider to mix in the sounds you want to play

**Comment by chikubian** [comment]
As I had a customer waiting on a fix, I've replaced my code with a "try to create and initialise a WaveOut player and if that fails, fallback to creating a DirectSoundOut instead" approach which makes it 100% reliable and only glitchy on the occasions where the WaveOut fails to initialise.

To answer your question, the crash occurred trying to allocate a second WaveOut instance, after a first one had been played, stopped and disposed, then a new first one was created and played, and while that one was playing, a second one was created and played.

Thanks for your suggestion with the MixingSampleProvider... I shall try this and get back to you. I'll probably still keep in the fallback to DirectSoundOut anyway, just in case those particular sound cards happen to ever fail when initialising the *first* WaveOut instance.

**Comment by chikubian** [comment]
In case it's useful, they're brand new Dell laptops with a "Realtek High Definition Audio" device showing in Device Manager, and the driver is Realtek's driver version 6.0.1.6088 published on Oct 8, 2015. Perhaps Realtek's quality assurance on drivers for their latest chips running on Windows 7 isn't so good now that Windows 7 is a legacy OS. 

Again, I appreciate the feedback you provided above. If there is any other information I can provide that would be useful, I'll have the misbehaving machines for the next few hours before I send them off to the customer.

**Comment by markheath** [comment]
I suppose one last thing you can try is going direct to Wasapi. Another thing might be to refactor your code so you only dispose a WaveOut once the PlaybackStopped event is raised. That's generally best practice.

**Comment by chikubian** [comment]
Thanks... I tried the refactoring you suggested (which makes a lot of sense) and it didn't help. If I have any significant problems with the DirectSoundOut fallback code, I'll try out Wasapi. Much appreciated Mark... thanks.
